File: backend/src/modules/msilo/msilo.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { Silo } from './entities/silo.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class MsiloService {
  constructor(
    @InjectRepository(Silo)
    private siloRepository: Repository<Silo>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  async findAll(dto: PaginationDto): Promise<PaginatedResult<Silo>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.username = Like(`%${keyword}%`);
    const [items, total] = await this.siloRepository.findAndCount({
      where, skip, take: limit, order: { incTime: 'DESC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async findByUser(username: string, domain: string): Promise<Silo[]> {
    return this.siloRepository.find({
      where: { username, domain },
      order: { incTime: 'DESC' },
    });
  }

  async remove(id: number): Promise<void> {
    const item = await this.siloRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('消息不存在');
    await this.siloRepository.remove(item);
  }

  async cleanExpired(): Promise<number> {
    const now = Math.floor(Date.now() / 1000);
    const result = await this.siloRepository
      .createQueryBuilder()
      .delete()
      .where('exp_time < :now', { now })
      .execute();
    return result.affected || 0;
  }

  async getStats(): Promise<any> {
    const total = await this.siloRepository.count();
    const now = Math.floor(Date.now() / 1000);
    const expired = await this.siloRepository
      .createQueryBuilder('s')
      .where('s.exp_time < :now', { now })
      .getCount();
    return { total, expired, active: total - expired };
  }

  async dump(): Promise<void> {
    await this.kamailioRpcService.call('msilo.dump');
  }
}

File: kam-admin-module-presence/src/main/java/io/kamailio/admin/modules/msilo/service/MsiloService.java
package io.kamailio.admin.modules.msilo.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.msilo.entity.Silo;
import io.kamailio.admin.modules.msilo.repository.SiloRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class MsiloService {
    private final SiloRepository repository;
    private final KamailioRpcService kamailioRpc;

    public MsiloService(SiloRepository repository, KamailioRpcService kamailioRpc) {
        this.repository = repository;
        this.kamailioRpc = kamailioRpc;
    }

    public PaginatedResult<Silo> findAll(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? repository.findByUsernameContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "incTime")))
                : repository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "incTime")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    public List<Silo> findByUser(String username, String domain) {
        return repository.findByUsernameAndDomainOrderByIncTimeDesc(username, domain);
    }

    @Transactional
    public void remove(Integer id) { repository.deleteById(id); }

    @Transactional
    public int cleanExpired() {
        int now = (int) (System.currentTimeMillis() / 1000);
        return repository.deleteByExpTimeBefore(now);
    }

    public Map<String, Object> getStats() {
        int now = (int) (System.currentTimeMillis() / 1000);
        long total = repository.count();
        long expired = repository.countByExpTimeBefore(now);
        Map<String, Object> result = new HashMap<>();
        result.put("total", total);
        result.put("expired", expired);
        result.put("active", total - expired);
        return result;
    }

    public void dump() { kamailioRpc.msiloDump().block(RpcTimeouts.DEFAULT_BLOCK); }
}
